storage/google/cloud/storage/blob.py

```diff
@@ -404,14 +408,22 @@ def _get_download_url(self):
         :returns: The download URL for the current blob.
         """
         if self.media_link is None:
-            download_url = _DOWNLOAD_URL_TEMPLATE.format(path=self.path)
+            base_url = _DOWNLOAD_URL_TEMPLATE.format(path=self.path)
+            scheme, netloc, path, query, frag = urlsplit(base_url)
+            query = parse_qsl(query)
+            # Only add `generation' for synthesized URLs: 'mediaLink' from
+            # server presumably already encodes it.
             if self.generation is not None:
-                download_url += u'&generation={:d}'.format(self.generation)
-            if self.user_project is not None:
-                download_url += u'&userProject={}'.format(self.user_project)
-            return download_url
+                query.append(('generation', '{:d}'.format(self.generation)))
         else:
-            return self.media_link
+            scheme, netloc, path, query, frag = urlsplit(self.media_link)
+            query = parse_qsl(query)
+
+        if self.user_project is not None:
+            query.append(('userProject', self.user_project))
```

Comment:
We can cross that bridge when we get there, but right now it's a weak argument. It's quite simple to just write code that does:

```python
if self.user_project is not None:
    download_url = _add_query_param(
        self.media_link, {'userProject': self.user_project})
else:
    download_url = self.media_link
```